File: backend/app/core/surveys.py
"""
=============================================================================
SURVEY DEFINITIONS
=============================================================================

Problem-validation surveys, defined in code so the server can reject any
answer that is not one of the offered options — the results are only worth
quoting if every row means exactly what the question asked.

Option ids are stable machine keys; the UI owns the uz/ru wording. Never
rename an id once responses exist — add a new one instead.
=============================================================================
"""

from typing import Any, Dict, List, Optional

# kind: "single" = one option id, "multi" = list of option ids, "text" = free text
SURVEYS: Dict[str, Dict[str, Any]] = {
    "student-2026": {
        "questions": [
            {"id": "status", "kind": "single", "required": True,
             "options": ["course_1_2", "course_3_4", "graduate_1y", "working", "other"]},
            {"id": "searched", "kind": "single", "required": True,
             "options": ["yes_found", "yes_searching", "no_plan", "no"]},
            {"id": "pains", "kind": "multi", "required": True, "max": 3,
             "options": ["experience_required", "hard_to_find", "scams", "resume",
                         "no_reply", "salary_unknown", "interview_prep", "other"]},
            {"id": "channels", "kind": "multi", "required": True,
             "options": ["telegram", "hh", "olx", "friends", "linkedin",
                         "university", "other"]},
            {"id": "time_to_job", "kind": "single", "required": False,
             "options": ["lt_1m", "m1_3", "m3_6", "gt_6m", "not_yet"]},
            {"id": "pay", "kind": "single", "required": True,
             "options": ["no", "k10_20", "k20_50", "k50_plus"]},
            {"id": "comment", "kind": "text", "required": False, "max_len": 1000},
        ],
    },
}

SOURCE_MAX_LEN = 40


def get_survey(key: str) -> Optional[Dict[str, Any]]:
    return SURVEYS.get(key)


def validate_answers(survey: Dict[str, Any], answers: Dict[str, Any]) -> Dict[str, Any]:
    """Return the cleaned answers, or raise ValueError naming the first bad field.

    Unknown keys are dropped rather than stored, so a tampered client can't
    park arbitrary JSON in the table.
    """
    if not isinstance(answers, dict):
        raise ValueError("answers must be an object")

    clean: Dict[str, Any] = {}
    for q in survey["questions"]:
        qid, kind = q["id"], q["kind"]
        value = answers.get(qid)
        empty = value is None or value == "" or value == []
        if empty:
            if q.get("required"):
                raise ValueError(f"{qid}: required")
            continue

        if kind == "single":
            if value not in q["options"]:
                raise ValueError(f"{qid}: unknown option")
            clean[qid] = value
        elif kind == "multi":
            if not isinstance(value, list) or not all(isinstance(v, str) for v in value):
                raise ValueError(f"{qid}: must be a list")
            picked: List[str] = list(dict.fromkeys(value))  # dedupe, keep order
            if any(v not in q["options"] for v in picked):
                raise ValueError(f"{qid}: unknown option")
            if len(picked) > q.get("max", len(q["options"])):
                raise ValueError(f"{qid}: too many options")
            clean[qid] = picked
        else:  # text
            if not isinstance(value, str):
                raise ValueError(f"{qid}: must be text")
            text = value.strip()[: q.get("max_len", 1000)]
            if text:
                clean[qid] = text
    return clean


def summarize(survey: Dict[str, Any], rows: List[Dict[str, Any]], text_limit: int = 100) -> Dict[str, Any]:
    """Per-option counts over `rows` (each a cleaned answers dict).

    For multi questions the percentages are of respondents, so they can sum
    past 100% — that is the honest reading of "pick up to three".
    """
    total = len(rows)
    questions = []
    for q in survey["questions"]:
        qid, kind = q["id"], q["kind"]
        if kind == "text":
            texts = [r[qid] for r in rows if r.get(qid)]
            questions.append({"id": qid, "kind": kind, "answered": len(texts),
                              "texts": texts[:text_limit]})
            continue
        counts = {opt: 0 for opt in q["options"]}
        answered = 0
        for r in rows:
            v = r.get(qid)
            if v is None:
                continue
            answered += 1
            for opt in (v if isinstance(v, list) else [v]):
                if opt in counts:
                    counts[opt] += 1
        questions.append({
            "id": qid, "kind": kind, "answered": answered,
            "options": [
                {"id": opt, "count": c,
                 "pct": round(100 * c / answered, 1) if answered else 0.0}
                for opt, c in counts.items()
            ],
        })
    return {"total": total, "questions": questions}

File: frontend/src/lib/surveys.ts
/**
 * Wording for the problem-validation survey. The question/option ids are the
 * server's (backend/app/core/surveys.py) — it rejects anything else, so keep
 * the two lists in step. Only the text lives here.
 */

export const STUDENT_SURVEY_KEY = "student-2026";

type L = { uz: string; ru: string };

export type SurveyQuestion = {
  id: string;
  kind: "single" | "multi" | "text";
  required: boolean;
  max?: number;
  title: L;
  hint?: L;
  options?: { id: string; label: L }[];
};

const o = (id: string, uz: string, ru: string) => ({ id, label: { uz, ru } });

export const STUDENT_SURVEY: SurveyQuestion[] = [
  {
    id: "status",
    kind: "single",
    required: true,
    title: { uz: "Hozir kimsiz?", ru: "Кто вы сейчас?" },
    options: [
      o("course_1_2", "1–2-kurs talabasi", "Студент 1–2 курса"),
      o("course_3_4", "3–4-kurs talabasi", "Студент 3–4 курса"),
      o("graduate_1y", "Bitirganimga 1 yil bo'lmagan", "Выпускник (до 1 года)"),
      o("working", "Ishlayapman", "Работаю"),
      o("other", "Boshqa", "Другое"),
    ],
  },
  {
    id: "searched",
    kind: "single",
    required: true,
    title: {
      uz: "Oxirgi 6 oyda ish yoki amaliyot qidirdingizmi?",
      ru: "Искали ли вы работу или стажировку за последние 6 месяцев?",
    },
    options: [
      o("yes_found", "Ha, topdim", "Да, нашёл(а)"),
      o("yes_searching", "Ha, hali qidiryapman", "Да, всё ещё ищу"),
      o("no_plan", "Yo'q, lekin yaqinda boshlayman", "Нет, но скоро начну"),
      o("no", "Yo'q", "Нет"),
    ],
  },
  {
    id: "pains",
    kind: "multi",
    required: true,
    max: 3,
    title: {
      uz: "Ish qidirishda eng katta qiyinchilik nima?",
      ru: "Что сложнее всего в поиске работы?",
    },
    hint: { uz: "3 tagacha tanlang", ru: "Выберите до 3" },
    options: [
      o("experience_required", "Tajriba talab qilishadi", "Требуют опыт"),
      o("hard_to_find", "Menga mos vakansiya topish qiyin", "Сложно найти подходящую вакансию"),
      o("scams", "Firibgar / soxta e'lonlar", "Мошеннические / фейковые объявления"),
      o("resume", "Rezyume yozishni bilmayman", "Не умею писать резюме"),
      o("no_reply", "Ish beruvchi javob bermaydi", "Работодатель не отвечает"),
      o("salary_unknown", "Maosh yozilmagan", "Не указана зарплата"),
      o("interview_prep", "Suhbatga tayyorlanish", "Подготовка к собеседованию"),
      o("other", "Boshqa", "Другое"),
    ],
  },
  {
    id: "channels",
    kind: "multi",
    required: true,
    title: { uz: "Ishni qayerdan qidirasiz?", ru: "Где вы ищете работу?" },
    hint: { uz: "Bir nechtasini tanlash mumkin", ru: "Можно выбрать несколько" },
    options: [
      o("telegram", "Telegram kanallar", "Telegram-каналы"),
      o("hh", "hh.uz", "hh.uz"),
      o("olx", "OLX", "OLX"),
      o("friends", "Tanishlar orqali", "Через знакомых"),
      o("linkedin", "LinkedIn", "LinkedIn"),
      o("university", "Universitet", "Университет"),
      o("other", "Boshqa", "Другое"),
    ],
  },
  {
    id: "time_to_job",
    kind: "single",
    required: false,
    title: {
      uz: "Oxirgi ishingizni topish qancha vaqt oldi?",
      ru: "Сколько времени занял поиск последней работы?",
    },
    options: [
      o("lt_1m", "1 oydan kam", "Меньше месяца"),
      o("m1_3", "1–3 oy", "1–3 месяца"),
      o("m3_6", "3–6 oy", "3–6 месяцев"),
      o("gt_6m", "6 oydan ko'p", "Больше 6 месяцев"),
      o("not_yet", "Hali topmadim", "Ещё не нашёл(а)"),
    ],
  },
  {
    id: "pay",
    kind: "single",
    required: true,
    title: {
      uz: "Tezroq ish topishga yordam beradigan xizmat uchun oyiga qancha to'lagan bo'lardingiz?",
      ru: "Сколько в месяц вы заплатили бы за сервис, который помогает быстрее найти работу?",
    },
    options: [
      o("no", "To'lamagan bo'lardim", "Не стал(а) бы платить"),
      o("k10_20", "10–20 ming so'm", "10–20 тыс. сум"),
      o("k20_50", "20–50 ming so'm", "20–50 тыс. сум"),
      o("k50_plus", "50 ming so'mdan ko'p", "Больше 50 тыс. сум"),
    ],
  },
  {
    id: "comment",
    kind: "text",
    required: false,
    title: {
      uz: "Ish qidirishda nima eng ko'p asabingizga tegadi? (ixtiyoriy)",
      ru: "Что больше всего раздражает в поиске работы? (необязательно)",
    },
  },
];

export type SurveyAnswers = Record<string, string | string[]>;

/** First required question left unanswered, or null when the form is complete. */
export function firstMissing(answers: SurveyAnswers, questions = STUDENT_SURVEY): string | null {
  for (const q of questions) {
    if (!q.required) continue;
    const v = answers[q.id];
    if (v === undefined || v === "" || (Array.isArray(v) && v.length === 0)) return q.id;
  }
  return null;
}

/** Toggle an option in a multi question, refusing to go past `max`. */
export function toggleMulti(current: string[] | undefined, id: string, max?: number): string[] {
  const list = current ?? [];
  if (list.includes(id)) return list.filter((x) => x !== id);
  if (max && list.length >= max) return list;
  return [...list, id];
}
